### ISSUE TITLE
Multiple server installations

### ISSUE THREAD

**Issue by profucius** [elblogbruno/NotionAI-MyMind#10]
I thought it would be best to split this question into a separate ticket.

>> So you are saying that I cannot install Notion AI server on the same computer multiple times with different ports? Does the server program itself conflict with other instances, even if the port is different?

> Sorry if I was misunderstood, yes you can run multiple servers with different ports, I believe, I haven't tried it, but in theory, it should work.

I am going to try this out, I am curious if it's possible for my own purposes. Currently I am trying to pinpoint where the server stores its configuration, so that I can try possibly containerizing it. I see that the server is default installed to C:\Program Files (x86)\Notion AI My Mind\  however where does the server.exe store its configuration, such as the port, credentials, etc?

**Comment by elblogbruno** [comment]
Nice investigation!
To answer your question it saves the credentials on the same folder where server.exe is located. Port.json tells the port to run the server on, data.json is where credentials are stored, and options.json is where the AI tagging options are saved.

**Comment by profucius** [comment]
I got it to work! muahahaha 😄

It was fairly easy to do, here are the steps I took:

1. I have a folder called D:\Portables which I store all portable software.
2. Inside this folder, I made a root folder called "Notion MyMind" (although it can be named anything)
3. Inside this folder, I made a folder and named it the topic that I am researching. (e.g. Investments)
3. Inside this folder, I copied C:\Program Files (x86)\Notion AI My Mind\ directly over.
4. So far, I have this folder: D:\Portables\Notion MyMind\Investments\Notion AI My Mind.
5. Note: I found that the final folder **must** be named exactly "Notion AI My Mind" because your .py script fails if it is not. However, any folder above this can be named anything the user desires.
6. Ran the server.exe.
7. Loaded the Firefox profile for my Investments research. Installed the Notion AI MyMind extension.
8. Configured the extension with a different port, and a different notion table. Same login credentials.
9. Saved text from a page, and it went straight into the Investments table within Notion!
10. Verified that this can be ran in parallel with other servers, by reproducing steps 3-7 above.
11. The second server successfully captures text into the different table.

While not ideal, I'm happy that this actually works. I have about 3 or 4 configs that I am going to setup like above. Of course, this is all in the meantime of one day being able to select tables on-the-fly within the extension. 😃

**Comment by elblogbruno** [comment]
Glad you got it working and thanks for sharing the steps, I will leave the issue open, so people can check into it.
I have updated to the 2.0.1 version, and my next work will be on adding the whole idea that we worked on in #8.

Thanks
Cordially,

**Comment by profucius** [comment]
Would it be possible to update your Windows installer to allow for selecting installation destination during setup? I just realized that when you update the installer, it defaults to Program Files (x86), which means I would need to copy the files out of that folder into each portable copy of the server. No rush if you're busy with other things.

**Comment by profucius** [comment]
Would it be possible to change the windows installer to ask for installation destination folder? This way, if a user is running multiple instances, they can update all of them by re-running the installer per folder.

**Comment by elblogbruno** [comment]
Well, I believe it would be easier and faster if I take the time to implement the original idea as I believe can't take long time.
![image](https://user-images.githubusercontent.com/10481058/110009420-48924080-7d1d-11eb-8d80-0a9027b60394.png)
Sneak peak 😮

**Comment by profucius** [comment]
Oh interesting, I'm excited! Okay, I will just sit tight on my idea then 😄

**Comment by elblogbruno** [comment]
In case you want more context on the photo, 
![image](https://user-images.githubusercontent.com/10481058/110010775-f520f200-7d1e-11eb-9c09-fa8c1ccf2090.png)
I am translating an structure like this to Json, that can be read from the extension and phone app.

**Comment by elblogbruno** [comment]
So you can add whatever collection of eleements and it will be translated in realtime 🤗 !

**Comment by profucius** [comment]
This looks really great!! I can't wait to see how it turns out 😄

**Comment by elblogbruno** [comment]
Me too 😄

**Comment by elblogbruno** [comment]
I have successfully created the functions needed on the server. I need to program the extension so it calls the API and gets the available collections and display them as HTML.

I will add them as buttons, so you will need to click twice, once for adding and second for choosing where to save. Maybe I could program if I don't click any to automatically add to a predetermined one.

Oh more ideas are coming here to my mind as I write this 😯 😂

**Comment by profucius** [comment]
Great ideas! This is sounding awesome already. Is that "click twice" on the context menu? Or is that like a popup type box?

**Comment by elblogbruno** [comment]
> Great ideas! This is sounding awesome already. Is that "click twice" on the context menu? Or is that like a popup type box?

Hi @profucius 
I am dealing with the chrome extension now. Somehow when page loads it makes a request and gets the available collections. So the idea is you click the button and the menu with buttons appears. Choose one and it will be added  to that collection.

**Comment by profucius** [comment]
Sounds like the way I was thinking it could work. Is it going well? The reason I ask is because there may be an alternative method you can consider. I've seen some other browser add-ons do this, and it may make it easier to update between Chrome and Firefox.

When the user clicks the Notion MyMind entry in the context menu, or on the browser bar, the extension opens a new tab (temporarily) to gather any other information (e.g. title, which table to capture to, etc) and when the user clicks a Save button on that tab, the tab closes itself and the server program continues processing.

This might be easier than programming a popup? 

If you've already figured out how to do the popup for Chrome and Firefox though, nevermind 😆

**Comment by elblogbruno** [comment]
I see the idea!
Currently extension is develop in some way the notification message pops up from all the context, when you add text / image or a website. 

Before , the popup only showed when adding a website as it was the official extension API, but I was not able to open this popup programatically speaking due to security reasons browser companies has established.

How it works now it injects html code on the website and this html can render the content. Sorry for all the technical 😂 this resumes in the idea I can show whatever content I add in html.

Eventough I love your idea, one of the things I want to always have on mind with this project is to be reliable and not very time consuming UI. That is why now you simply click to add instead of having to think more about what you are doing.

I am very glad for all the feedback, I definitely have it in my radar as how to make the UI and characteristics get to next level.

Speaking about if it is going well, I expect to have it this Monday or Tuesday, or this week we can say.

**Comment by profucius** [comment]
That makes sense! I follow what you're saying. I like that a lot, keeping "how" you use it simple, while still being able to add features. That is good software development protocol.

I can't wait to check it out whenever it is ready. I'll be downloading it that day 😁

**Comment by elblogbruno** [comment]
![image](https://user-images.githubusercontent.com/10481058/110247622-69e96b80-7f6d-11eb-86e4-b91784c42ce8.png)
Sneak peek

**Comment by elblogbruno** [comment]
@profucius you can take a look now at latest release. Hope you like it! 
If you like it, you may close this issue!

**Comment by elblogbruno** [comment]
I am updating the android app, expect a release today! 😁

**Comment by elblogbruno** [comment]
Just release the android and extensions! If you feel like you like it, please close this issue!
As always. it will take some time for google to publish the updates on the chrome web store and play store.
